File: 백준/Silver/10844. 쉬운 계단 수/쉬운 계단 수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

// 10844. 쉬운 계단 수 / 실버1 / 2:16 ~ 2:23
public class Main {
    static final int MOD = 1_000_000_000;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        long[][] dp = new long[n + 1][10];
        for (int i = 1; i <= 9; i++) {
            dp[1][i] = 1;
        }

        for (int i = 2; i <= n; i++) {
            dp[i][0] = dp[i - 1][1] % MOD;
            dp[i][9] = dp[i - 1][8] % MOD;
            for (int j = 1; j <= 8; j++) {
                dp[i][j] = (dp[i - 1][j - 1] + dp[i - 1][j + 1]) % MOD;
            }
        }

        long sum = 0;
        for (int i = 0; i <= 9; i++) {
            sum += dp[n][i];
            sum %= MOD;
        }

        System.out.println(sum);
    }
}

File: 백준/Silver/10844. 쉬운 계단 수/쉬운 계단 수.cc
#include<iostream>
using namespace std;

long long int d[101][10];

int main()
{
	int n;
	cin >> n;
	for (int i = 1; i <= 9; i++)
		d[1][i] = 1;

	for (int i = 2; i <= n; i++)
	{
		d[i][0] = d[i - 1][1] % 1000000000;
		d[i][9] = d[i - 1][8] % 1000000000;
		for (int j = 1; j <= 8; j++)
			d[i][j] = (d[i - 1][j - 1] + d[i - 1][j + 1]) % 1000000000;
	}

	long long int sum = 0;
	for (int i = 0; i <= 9; i++)
		sum += d[n][i] % 1000000000;
	cout << sum % 1000000000;

	return 0;
}